webcam-capture-drivers/driver-ffmpeg-cli/src/main/java/com/github/sarxos/webcam/ds/ffmpegcli/FFmpegCliDevice.java

```diff
@@ -1,260 +1,218 @@
-						break; // EOI found
-					}
-				}
-			} while (true);
-
-		} catch (IOException e) {
-			throw new RuntimeException(e);
-		}
-
-		return baos.toByteArray();
-
-	}
-
-	@Override
-	public synchronized ByteBuffer getImageBytes() {
-		if (!open.get()) {
-			return null;
-		}
-		return ByteBuffer.wrap(readBytes());
-	}
-
-	@Override
-	public void getImageBytes(ByteBuffer buffer) {
-		if (!open.get()) {
-			return;
-		}
-		buffer.put(readBytes());
-	}
-
-	@Override
-	public BufferedImage getImage() {
-
-		if (!open.get()) {
-			return null;
-		}
-
-		ByteArrayInputStream bais = new ByteArrayInputStream(readBytes());
-		try {
-			return ImageIO.read(bais);
-		} catch (IOException e) {
-			throw new RuntimeException(e);
-		} finally {
-			try {
-				bais.close();
-			} catch (IOException e) {
-				throw new RuntimeException(e);
-			}
-		}
-	}
-
-	@Override
-	public synchronized void open() {
-
-		if (disposed.get()) {
-			return;
-		}
-
-		if (!open.compareAndSet(false, true)) {
-			return;
-		}
-
-		pipe = new File("/tmp/" + vfile.getName() + ".mjpeg");
-
-		//@formatter:off
-		String[] cmd = new String[] { 
-			"ffmpeg", 
-			"-y",                          // overwrite output file 
-			"-f", "video4linux2",          // format
-			"-input_format", "mjpeg",      // input format
-			"-r", "50",                    // requested FPS
-			"-s", getResolutionString(),   // frame dimension
-			"-i", vfile.getAbsolutePath(), // input file 
-			"-vcodec", "copy",             // video codec to be used
-			pipe.getAbsolutePath(),        // output file
-		};
-		//@formatter:on
-
-		if (LOG.isDebugEnabled()) {
-			StringBuilder sb = new StringBuilder();
-			for (String c : cmd) {
-				sb.append(c).append(' ');
-			}
-			LOG.debug("Executing command: {}", sb.toString());
-		}
-
-		try {
-			RT.exec(new String[] { "mkfifo", pipe.getAbsolutePath() }).waitFor();
-			process = RT.exec(cmd);
-			dis = new DataInputStream(new FileInputStream(pipe));
-		} catch (FileNotFoundException e) {
-			throw new RuntimeException(e);
-		} catch (IOException e) {
-			throw new RuntimeException(e);
-		} catch (InterruptedException e) {
-			throw new RuntimeException(e);
-		}
-	}
-
-	@Override
-	public synchronized void close() {
-
-		if (!open.compareAndSet(true, false)) {
-			return;
-		}
-
-		try {
-			dis.close();
-		} catch (IOException e) {
-			throw new RuntimeException(e);
-		}
-
-		process.destroy();
-
-		try {
-			process.waitFor();
-		} catch (InterruptedException e) {
-			throw new RuntimeException(e);
-		}
-
-		if (!pipe.delete()) {
-			pipe.deleteOnExit();
-		}
-	}
-
-	@Override
-	public void dispose() {
-		if (disposed.compareAndSet(false, true) && open.get()) {
-			close();
-		}
-	}
-
-	@Override
-	public boolean isOpen() {
-		return open.get();
-	}
-
-	@Override
-	public String toString() {
-		return "video device " + name;
-	}
+    private static final Logger LOG = LoggerFactory.getLogger(FFmpegCliDevice.class);
+
+    private Process process = null;
```

Comment:
Process is set/get from a different threads and therefore should be marked with `volatile` keyword. Otherwise you may end up with some weird side-effects, e.g. when you read `process` it may be regardless the fact that other thread already set it to non-null. The other side effect is that you may get old instance when running this twice.

```java
private volatile Process process = null;
```

http://tutorials.jenkov.com/java-concurrency/volatile.html#volatile-is-not-always-enough